### PR TITLE
enable sticky sessions

### PR DESCRIPTION
**Pull request by mitchdraft**
BOT NOTES: 
resolves https://github.com/solo-io/gloo/issues/935

### CODE REVIEW

**Comment by mitchdraft** [comment]
wip while adding tests

File: projects/gloo/api/v1/load_balancer.proto [hunk 1]
```diff
@@ -30,13 +32,38 @@ message LoadBalancerConfig {
...
+    }
+
+    message RingHash {
+        // Optional, customizes the parameters used in the hashing algorithm
+        RingHashConfig ring_hash_config = 1;
+    }
+
+    message Maglev {
+        // Optional, customizes the parameters used in the hashing algorithm
+        RingHashConfig ring_hash_config = 1;
```
Comment (EItanya) [inline projects/gloo/api/v1/load_balancer.proto]: According to the docs, maglev doesn't have additional config currently
https://github.com/envoyproxy/envoy/blob/e349fb6139e4b7a59a9a359be0ea45dd61e589c5/api/envoy/api/v2/cds.proto#L500

File: projects/gloo/pkg/plugins/loadbalancer/plugin.go [hunk 2]
```diff
@@ -20,10 +32,54 @@ func (p *Plugin) Init(params plugins.InitParams) error {
...
+	if err := utils.EnsureRouteAction(out); err != nil {
+		return InvalidRouteTypeError(err)
+	}
+	outRa := out.GetRoute()
+	outRa.HashPolicy = getHashPoliciesFromSpec(lbPlugin.HashPolicies)
 	return nil
 }
 
+func getHashPoliciesFromSpec(spec []*lbhash.HashPolicy) []*envoyroute.RouteAction_HashPolicy {
+	if len(spec) == 0 {
```
Comment (EItanya) [inline projects/gloo/pkg/plugins/loadbalancer/plugin.go]: unecessary check, the for loop will skip and nil will be returned

File: projects/gloo/pkg/plugins/loadbalancer/plugin.go [hunk 3]
```diff
@@ -60,8 +116,34 @@ func (p *Plugin) ProcessUpstream(params plugins.Params, in *v1.Upstream, out *en
 			}
 		case *v1.LoadBalancerConfig_Random_:
 			out.LbPolicy = envoyapi.Cluster_RANDOM
+		case *v1.LoadBalancerConfig_RingHash_:
+			out.LbPolicy = envoyapi.Cluster_RING_HASH
+			setRingHashLbConfig(out, lbtype.RingHash.RingHashConfig)
+		case *v1.LoadBalancerConfig_Maglev_:
+			out.LbPolicy = envoyapi.Cluster_MAGLEV
+			setRingHashLbConfig(out, lbtype.Maglev.RingHashConfig)
```
Comment (EItanya) [inline projects/gloo/pkg/plugins/loadbalancer/plugin.go]: Same comment as above about maglev config

File: projects/gloo/api/v1/load_balancer.proto (hunk 1 above)
Comment (mitchdraft) [reply projects/gloo/api/v1/load_balancer.proto]: You're right, I took the "maglev is a drop in for ring hash" statement too literally.
https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/upstream/load_balancing/load_balancers.html?highlight=maglev

I tested with and without the ring hash config, makes no difference, so I removed it.
Good catch

**Comment by mitchdraft** [comment]
> note that you also want to update
> projects/gloo/pkg/plugins/utils/update_upstream_spec.go
> so that discovery doesn't override your changes.

@yuval-k I think this is already done here:
https://github.com/solo-io/gloo/blob/master/projects/gloo/pkg/plugins/utils/update_upstream_spec.go#L18

**Pull request merged**